plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetescluster/KubernetesClusterManagerImpl.java

```diff
@@ -0,0 +1,3105 @@
+
+        Field cidrField = rule.getClass().getDeclaredField("cidrlist");
+        cidrField.setAccessible(true);
+        cidrField.set(rule, sourceCidrList);
+
+        firewallService.createIngressFirewallRule(rule);
+        firewallService.applyIngressFwRules(publicIp.getId(), account);
+    }
+
+    private void removePortForwardingRules(IpAddress publicIp, Network network, Account account, List<Long> removedVMIds) throws ResourceUnavailableException {
+        if (!CollectionUtils.isEmpty(removedVMIds)) {
+            for (Long vmId : removedVMIds) {
+                List<PortForwardingRuleVO> pfRules = portForwardingRulesDao.listByNetwork(network.getId());
+                for (PortForwardingRuleVO pfRule : pfRules) {
+                    if (pfRule.getVirtualMachineId() == vmId) {
+                        portForwardingRulesDao.remove(pfRule.getId());
+                        break;
+                    }
+                }
+            }
+            rulesService.applyPortForwardingRules(publicIp.getId(), account);
+        }
+    }
+
+    private void provisionSshPortForwardingRules(KubernetesCluster kubernetesCluster, IpAddress publicIp, Network network, Account account, List<Long> clusterVMIds, int firewallRuleSourcePortStart) throws ResourceUnavailableException,
+            NetworkRuleConflictException {
+        if (!CollectionUtils.isEmpty(clusterVMIds)) { // Upscaling, add new port-forwarding rules
+            // Apply port forwarding only to new VMs
+            final long publicIpId = publicIp.getId();
+            final long networkId = network.getId();
+            final long accountId = account.getId();
+            final long domainId = account.getDomainId();
+            for (int i = 0; i < clusterVMIds.size(); ++i) {
+                long vmId = clusterVMIds.get(i);
+                Nic vmNic = networkModel.getNicInNetwork(vmId, networkId);
+                final Ip vmIp = new Ip(vmNic.getIPv4Address());
+                final long vmIdFinal = vmId;
+                final int srcPortFinal = firewallRuleSourcePortStart + i;
+
+                PortForwardingRuleVO pfRule = Transaction.execute(new TransactionCallbackWithException<PortForwardingRuleVO, NetworkRuleConflictException>() {
+                    @Override
+                    public PortForwardingRuleVO doInTransaction(TransactionStatus status) throws NetworkRuleConflictException {
+                        PortForwardingRuleVO newRule =
+                                new PortForwardingRuleVO(null, publicIpId,
+                                        srcPortFinal, srcPortFinal,
+                                        vmIp,
+                                        22, 22,
+                                        "tcp", networkId, accountId, domainId, vmIdFinal);
+                        newRule.setDisplay(true);
+                        newRule.setState(FirewallRule.State.Add);
+                        newRule = portForwardingRulesDao.persist(ne
```

Comment:
the comment can be removed now, the method call is actually more descriptive, very nice 👍